planner/core/planbuilder_test.go

```diff
@@ -172,3 +185,317 @@ func (s *testPlanBuilderSuite) TestDisableFold(c *C) {
+		AggFuncs:     aggDescs,
+		GroupByItems: []expression.Expression{col, cst},
+	}}
+	streamAgg = streamAgg.initForStream(ctx, stats, 0)
+	streamAgg.SetSchema(schema)
+	c.Assert(checkPhysicalPlanClone(streamAgg), IsNil)
+
+	// hash agg
+	hashAgg := &PhysicalHashAgg{basePhysicalAgg{
+		AggFuncs:     aggDescs,
+		GroupByItems: []expression.Expression{col, cst},
+	}}
+	hashAgg = hashAgg.initForHash(ctx, stats, 0)
+	hashAgg.SetSchema(schema)
+	c.Assert(checkPhysicalPlanClone(hashAgg), IsNil)
+
+	// hash join
+	hashJoin := &PhysicalHashJoin{
+		Concurrency:     4,
+		UseOuterToBuild: true,
+	}
+	hashJoin = hashJoin.Init(ctx, stats, 0)
+	hashJoin.SetSchema(schema)
+	c.Assert(checkPhysicalPlanClone(hashJoin), IsNil)
+
+	// merge join
+	mergeJoin := &PhysicalMergeJoin{
+		CompareFuncs: []expression.CompareFunc{expression.CompareInt},
+		Desc:         true,
+	}
+	mergeJoin = mergeJoin.Init(ctx, stats, 0)
+	mergeJoin.SetSchema(schema)
+	c.Assert(checkPhysicalPlanClone(mergeJoin), IsNil)
+}
+
+//go:linkname valueInterface reflect.valueInterface
+func valueInterface(v reflect.Value, safe bool) interface{}
+
+func typeName(t reflect.Type) string {
+	path := t.String()
+	tmp := strings.Split(path, ".")
+	if len(tmp) == 1 {
+		return tmp[0]
+	}
+	return tmp[len(tmp)-1]
+}
+
+func checkPhysicalPlanClone(p PhysicalPlan) error {
+	cloned, err := p.Clone()
+	if err != nil {
+		return err
+	}
+	whiteList := []string{"*property.StatsInfo", "*sessionctx.Context", "*mock.Context", "*types.FieldType"}
+	return checkDeepCloned(reflect.ValueOf(p), reflect.ValueOf(cloned), typeName(reflect.TypeOf(p)), whiteList, nil)
+}
+
+// checkDeepCloned is used to check if v2 is deep cloned from v1.
+// It's modified from reflect.deepValueEqual.
+func checkDeepCloned(v1, v2 reflect.Value, path string, whiteList []string, visited map[visit]bool) error {
+	if !v1.IsValid() || !v2.IsValid() {
+		if v1.IsValid() != v2.IsValid() {
+			return errors.Errorf("invalid")
+		}
+		return nil
+	}
+	if v1.Type() != v2.Type() {
+		return errors.Errorf("different type %v, %v, path %v", v1.Type(), v2.Type(), path)
+	}
+
+	if visited == nil {
+		visited = make(map[visit]bool)
+	}
+	hard := func(k reflect.Kind) bool {
+		switch k {
+		case reflect.Map, reflect.Slice, reflect.Ptr, reflect.Interface:
+			return true
+		}
+		return false
+	}
+	if v1.CanAddr() && v2.CanAddr() && hard(v1.Kind()) {
+		addr1 := unsafe.Pointer(v1.UnsafeAddr())
+		addr2 := unsafe.Pointer(v2.UnsafeAddr())
+		if uintptr(addr1) > uintptr(addr2) {
+			addr1, addr2 = addr2, addr1
+		}
+		typ := v1.Type()
+		v := visit{addr1, addr2, typ}
+		if visited[v] {
+			return nil
+		}
+		visited[v] = true
+	}
+
+	switch v1.Kind() {
+	case reflect.Array:
+		for i := 0; i < v1.Len(); i++ {
+			if err := checkDeepCloned(v1.Index(i), v2.Index(i), fmt.Sprintf("%v[%v]", path, i), whiteList, visited); err != nil {
+				return err
+			}
+		}
+	case reflect.Slice:
+		if v1.IsNil() && v2.IsNil() || (v1.Len() == 0 && v2.Len() == 0) {
```

Comment:
When `v1.IsNil() = true && v2.IsNil() != true`, will it panic here?

Reply:
updated.